### ISSUE TITLE
Convert Python file object to a Rust file object

### ISSUE THREAD

**Issue by thedrow** [PyO3/pyo3#933]
When dealing with files, it'd be easier to use the Rust file API whenever a file object is passed.
For that matter, we need a PyFile type as well.

**Comment by davidhewitt** [comment]
Note that creating Rust files from a file descriptor is unsafe because it assumes ownership of the file descriptor, which wouldn't be the case here (the Python file object would continue to own the descriptor). Probably to do it correctly we'd need to call something like `libc::dup()` on the file descriptor. Also I guess something similar on Windows.

**Comment by althonos** [comment]
Not sure how safe it is, but you hqve this: https://docs.rs/pyo3-file/0.3.1/pyo3_file/struct.PyFileLikeObject.html

**Comment by ethanhs** [comment]
This will be annoying on Windows. Windows (and Rust) use handles to describe files, but Python seems to use the Windows compatibility layer around file descriptors.

**Comment by thedrow** [comment]
> Not sure how safe it is, but you have this: https://docs.rs/pyo3-file/0.3.1/pyo3_file/struct.PyFileLikeObject.html

That's cool but maybe we can do better than that?

**Comment by thedrow** [comment]
> This will be annoying on Windows. Windows (and Rust) use handles to describe files, but Python seems to use the Windows compatibility layer around file descriptors.

What would you like to do? Do you have a plan to deal with this issue?

**Comment by ethanhs** [comment]
@thedrow I'm not really sure what the best path is for Windows. There is an API to convert file descriptors to handles, but I think it will make lifetimes difficult: https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/get-osfhandle?view=vs-2019

**Comment by thedrow** [comment]
I know absolutely nothing about Windows.
So either we figure this out or keep this issue open until we do.

**Comment by ethanhs** [comment]
Actually, it looks like [`FromRawHandle`](https://doc.rust-lang.org/std/os/windows/io/trait.FromRawHandle.html) requires ownership of a handle, which seems impossible to make work since Python's `io.BaseIO.fileno` is a file descriptor. :/

**Comment by thedrow** [comment]
So the problem is in Rust? Should we take it upstream?

**Comment by ethanhs** [comment]
Yeah it seems that it may be good to have an API (there may be one I'm unaware of) to interface with `FILE*` descriptors on Windows with Rust.

**Comment by deedy5** [comment]
Any news on this issue?

**Comment by alex** [comment]
This issue is up to date with the current state of this work (i.e., no recent updates).
